### DISCUSSION TITLE
Idea for millis: Combine two 16-Bit timers for a 32-bit timer (no ISR)

### DISCUSSION THREAD

**Discussion by MX682X** [SpenceKonde/DxCore#347] [Future development discussion]
I have no idea how complicated it might end up, but I had an idea for ISR free millis (one possiblity would be timekeeping in STAND-BY without CPU wakeup) 
e.g. since available on all chips: RTC fires Event on overflow and TCBx counts it. When calling millis, we just read the registers of those Timers to get the time

**Comment by SpenceKonde** [comment]
Yeah - I've thought about this too. The annoying thing is that you need more than two timers, if you want to do millis with it and count to the same number, because you would need to get the TCB running at 1 ms per tick, which would mean you would need a third timer to trigger the two cascaded timers, or render a TCA unavailable for PWM. That was what kinda sunk that idea for me. I think the much more useful thing would be way to use the RTC to keep time while sleeping and then switch over to a higher speed timer when awake. I put in a lot of the infrastructure for it - my hope was to make "sleeplib" to do things like that. and automatically return to sleep if the wakeup conditions weren't met. 
I don't know that there's much of a demand for an ISR-less millis timer? Unless it were based on the RTC to do it in sleep, but then you lose micros unless you switch to a highspeed timer while awake. And you can't do power saving if you're using it without an ISR at the RTC overflow because TCB requires the clock to be running, and leaving the high speed clock on defeats the point of power save in sleep. What one really wants for this is an event triggered counter that works fully async, and hence doesn't require keeping the high speed oscillator on... But we don't have one of those

**Reply by MX682X** [reply]
yeah, I've realized later that the RTC can't count milliseconds without extra math.  Wanted to delete this discussion then, but didn#t find a button to do so....
But I am thinking about suggesting Microchip to add a fuse or so that lets the user decide to clock the ULP oscillator at 32.000kHz.
And maybe a 32-Bit event edge counter...

**Reply by SpenceKonde** [reply]
The RTC can count milliseconds without math if you're willing to use an external 32.000 kHz crystal, just not a 32.768 kHz

**Reply by SpenceKonde** [reply]
Obviously you lose autotune.
